### PR TITLE
chore: group-by primitive key support

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Fixes: https://github.com/confluentinc/ksql/issues/4092

This commit gets `GROUP BY` clauses working with primitive key types.

BREAKING CHANGE: A `GROUP BY` on single expressions now changes the SQL type of `ROWKEY` in the output schema of the query to match the SQL type of the expression.

 For example, consider:

 ```sql
 CREATE STREAM INPUT (ROWKEY STRING KEY, ID INT) WITH (...);
 CREATE TABLE OUTPUT AS SELECT COUNT(*) AS COUNT FROM INPUT GROUP BY ID;
 ```

Previously, the above would have resulted in an output schema of `ROWKEY STRING KEY, COUNT BIGINT`, where `ROWKEY` would have stored the string representation of the integer from the `ID` column.

With this commit the output schema will be `ROWKEY INT KEY COUNT BIGINT`.

BREAKING CHANGE: Any`GROUP BY` expression that resolves to `NULL`, including because a UDF throws an exception, now results in the row being excluded from the result.  Previously, as the key was a `STRING` a value of `"null"` could be used. With other primitive types this is not possible. As key columns must be non-null any exception is logged and the row is excluded.

Note: I've had to change how QTT determines which key serde to use to support checking internal topic messages.  This PR contains the minimum changes needed to get the tests passing. I'll clean up the rest in a follow up PR.

### Testing done 

Suitable tests added to QTT.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-streams/src/main/java/io/confluent/ksql/execution/streams/GroupByParamsFactory.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  public static GroupByParams build(
+      final LogicalSchema sourceSchema,
+      final List<ExpressionMetadata> expressions
+  ) {
+    final Function<GenericRow, Struct> mapper = expressions.size() == 1
+        ? new SingleExpressionGrouper(expressions.get(0))::apply
+        : new MultiExpressionGrouper(expressions)::apply;
+
+    final LogicalSchema schema = expressions.size() == 1
+        ? singleExpressionSchema(sourceSchema, expressions.get(0).getExpressionType())
```
Comment (rodesai) [inline ksql-streams/src/main/java/io/confluent/ksql/execution/streams/GroupByParamsFactory.java]: Where do we validate that this type is a primitive?

File: ksql-streams/src/main/java/io/confluent/ksql/execution/streams/StepSchemaResolver.java [hunk 2]
```diff
@@ -160,6 +162,22 @@ private LogicalSchema handleStreamFlatMap(
     );
   }
 
+  private LogicalSchema handleGroupBy(
+      final LogicalSchema sourceSchema,
+      final StreamGroupBy<?> streamGroupBy
+  ) {
+    final List<Expression> groupBy = streamGroupBy.getGroupByExpressions();
+
```
Comment (rodesai) [inline ksql-streams/src/main/java/io/confluent/ksql/execution/streams/StepSchemaResolver.java]: Can't this just be:
`return GroupByParamsFactory.build(sourceSchema, streamGroupBy.getGroupByExpressions().getSchema();`

File: ksql-streams/src/main/java/io/confluent/ksql/execution/streams/GroupByParamsFactory.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-streams/src/main/java/io/confluent/ksql/execution/streams/GroupByParamsFactory.java]: I'll be checking that all works nicely in https://github.com/confluentinc/ksql/issues/4113

File: ksql-streams/src/main/java/io/confluent/ksql/execution/streams/StepSchemaResolver.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-streams/src/main/java/io/confluent/ksql/execution/streams/StepSchemaResolver.java]: Nope. We'd need to first compile the expressions into ExpressionMetadata, which seems a little unnecessary / costly.

Though avoiding that could be a case of premature optimisation... thoughts?

**Pull request merged**